src/toil/wdl/toilwdl.py

```diff
@@ -0,0 +1,2125 @@
+
+    def parse_workflow(self, workflow):
+        '''
+        Parses a WDL workflow AST subtree.
+
+        Currently looks at and parses 3 sections:
+        1. Declarations (e.g. string x = 'helloworld')
+        2. Calls (similar to a python def)
+        3. Scatter (which expects to map to a Call or multiple Calls)
+
+        Returns nothing but creates the self.workflows_dictionary necessary for much
+        of the parser.
+
+        :param workflow: An AST subtree of a WDL "Workflow".
+        :return: Returns nothing but adds a workflow to the
+                 self.workflows_dictionary necessary for much of the parser.
+        '''
+        workflow_name = workflow.attr('name').source_string
+
+        wf_declared_dict = {}
+        for section in workflow.attr("body"):
+
+            if section.name == "Declaration":
+                var_name, var_map = self.parse_workflow_declaration(section)
+                wf_declared_dict[var_name] = var_map
+            self.workflows_dictionary.setdefault(workflow_name, {})['wf_declarations'] = wf_declared_dict
+
+            if section.name == "Scatter":
+                self.task_priority = self.task_priority + 1
+
+                # name of iterator; e.g. 'sample'
+                # also serves as a variable input in function for indexed variables; e.g. sample[0], sample[1], etc.
+                scatter_counter = section.attributes['item'].source_string
+                # name of collection to iterate over
+                scatter_collection = section.attributes['collection'].source_string
+
+                self.workflows_dictionary.setdefault('scatter_calls', {})[scatter_collection] = scatter_counter
+
+                if scatter_collection in self.workflows_dictionary[workflow_name]['wf_declarations']:
+                    if self.workflows_dictionary[workflow_name]['wf_declarations'][scatter_collection]['type'] == 'Array':
+                        scatter_array = self.workflows_dictionary[workflow_name]['wf_declarations'][scatter_collection]['value']
+                        scatter_num = 0
+                        for set_of_vars in scatter_array:
+                            for j in section.attributes['body']:
+                                self.task_number = self.task_number + 1
+                                task_being_called = j.attributes['task'].source_string
+                                if j.attributes['alias']:
+                                    task_alias = j.attributes['alias'].source_string
+                                else:
+                                    task_alias = task_being_called
+                                job = self.parse_workflow_call(j, scatter_num=str(scatter_num))
+                                self.workflows_dictionary.setdefault((self.task_priority, self.task_number, task_being_called, task_alias), {})['job_declarations'] = job
+                            scatter_num = scatter_num + 1
+                    else:
```

Comment:
this code block is too big for those of use with self-inflicted brain damage to understand.  Again suggest, small well named functions make code easier to understand

Reply:
Thanks Mark!  I broke up those two functions in an effort to save myself from code-induced brain damage in the future.  It's a bad habit that I haven't completely gotten rid of, and thanks for pointing it out!